### PR TITLE
[SPARK-30119][WEBUI] Add Pagination Support to Streaming Page

### PR DESCRIPTION
**Pull request by iRakson**
### What changes were proposed in this pull request?
* Pagination Support is added to all tables of streaming page in spark web UI.
For adding pagination support, existing classes from #7399 were used.
* Earlier streaming page has two tables `Active Batches` and `Completed Batches`. Now, we will have three tables `Running Batches`, `Waiting Batches` and `Completed Batches`. If we have large number of waiting and running batches then keeping track in a single table is difficult. Also other pages have different table for different type type of data.
* Earlier empty tables were shown. Now only non-empty tables will be shown.
`Active Batches` table used to show details of waiting batches followed by running batches. 

### Why are the changes needed?
Pagination will allow users to analyse the table in much better way. All spark web UI pages support pagination apart from streaming pages, so this will add consistency as well. Also it might fix the potential OOM errors that can arise. 

### Does this PR introduce _any_ user-facing change?
Yes. `Active Batches` table is split into two tables `Running Batches` and `Waiting Batches`. Pagination Support is added to the all the tables. Every other functionality is unchanged.

### How was this patch tested?
UT added.

Before changes:
Screenshot 2020-05-03 at 7 07 14 PM

After Changes:
Screenshot 2020-06-01 at 11 26 48 PM

### CODE REVIEW

**Comment by iRakson** [comment]
cc @srowen @dongjoon-hyun @HyukjinKwon Kindly review

**Comment by srowen** [comment]
Jenkins test this please

**Comment by iRakson** [comment]
retest this please

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Should https://github.com/apache/spark/pull/28448 come before this?
I think we do want pagination here and trust your changes, but don't know this code well.
Not sure if @sarutak or @zsxwing wants to take a look.

**Comment by iRakson** [comment]
> Should #28448 come before this?

Yeah. I think it would be better if we first clean up.

**Comment by iRakson** [comment]
@srowen @sarutak PR updated with latest changes in Pagination Framework.

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Does this need to go in before or after https://github.com/apache/spark/pull/28448 ? want to make sure I understand the desired order and dependencies.

**Comment by iRakson** [comment]
Both are independent of each other. Either one can go first.
#28448 only cleans code.

**Comment by iRakson** [comment]
@sarutak Kindly take a look at this one.

**Comment by srowen** [comment]
Jenkins test this please

**Comment by sarutak** [comment]
@iRakson Sorry for the late reply. I'll check this weekend.

**Comment by iRakson** [comment]
It's ok. Please check according to your convenience. :)

**Review by sarutak — COMMENTED** [review]
@iRakson This change newly adds paged tables so I think it's better to add test cases too to test whether this change uses the pagination frame framework correctly.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/AllBatchesTable.scala [hunk 1]
```diff
@@ -49,147 +63,152 @@ private[ui] abstract class BatchTableBase(tableId: String, batchInterval: Long)
...
+  override def tableId: String = s"$tableTag-table"
 
-  def toNodeSeq: Seq[Node] = {
-    batchTable
-  }
+  override def tableCssClass: String =
+    "table table-bordered table-sm table-striped table-head-clickable table-cell-width-limited"
 
-  protected def createOutputOperationProgressBar(batch: BatchUIData): Seq[Node] = {
-    <td class="progress-cell">
```
[the remark this answers is not in the document]
Comment (iRakson) [reply streaming/src/main/scala/org/apache/spark/streaming/ui/AllBatchesTable.scala:93]: I removed to maintain consistency. i.e 
```
<td> ...... </td>
<td> ...... </td>
```

Anyway, i will use the original code.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/AllBatchesTable.scala [hunk 2]
```diff
@@ -49,147 +63,152 @@ private[ui] abstract class BatchTableBase(tableId: String, batchInterval: Long)
...
+  override def headers: Seq[Node] = {
+    // tuple containing tooltips for header fields
+    val tooltips = ("Time taken by Streaming scheduler to submit jobs of a batch",
+      "Time taken to process all jobs of a batch", "Total time taken to handle a batch")
+    // headers, sortable and tooltips
+    val headersAndCssClasses: Seq[(String, Boolean, Option[String])] = {
+      Seq(
+        ("Batch Time", true, None),
+        ("Records", true, None),
+        ("Scheduling Delay", true, Some(tooltips._1)),
```
[the remark this answers is not in the document]
Comment (iRakson) [reply streaming/src/main/scala/org/apache/spark/streaming/ui/AllBatchesTable.scala:109]: I worked on this before working on other pages, i missed this one. I will update to literals.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/AllBatchesTable.scala [hunk 3]
```diff
@@ -17,30 +17,44 @@
...
+
+private[ui] class StreamingPagedTable(
+    request: HttpServletRequest,
+    tableTag: String,
+    batches: Seq[BatchUIData],
+    basePath: String,
+    subPath: String,
+    isRunningTable: Boolean,
+    isWaitingTable: Boolean,
+    isCompletedTable: Boolean,
```
[the remark this answers is not in the document]
Comment (iRakson) [reply streaming/src/main/scala/org/apache/spark/streaming/ui/AllBatchesTable.scala:36]: Thank you suggesting this. I will change to constants.

**Comment by iRakson** [comment]
@sarutak Thank you for reviewing. I will update PR with test cases and suggested code changes.

**Comment by iRakson** [comment]
@sarutak @srowen Kindly take a look.
I added few tests for checking number of rows in paginated tables, sorting.

**Comment by iRakson** [comment]
retest this please

**Comment by sarutak** [comment]
add to whitelist.

**Comment by sarutak** [comment]
@iRakson From now on, Jenkins starts on pushing your commit and you can start Jenkins by saying "retest this please".

**Comment by iRakson** [comment]
> @iRakson From now on, Jenkins starts on pushing your commit and you can start Jenkins by saying "retest this please".

@sarutak Thank you for adding me to whitelist :)

**Comment by iRakson** [comment]
retest this please

**Comment by iRakson** [comment]
@sarutak Gentle ping.

**Comment by srowen** [comment]
If you're OK with it @sarutak I'll merge

**Comment by sarutak** [comment]
I've inspected again and it seems ok to me.

**Comment by sarutak** [comment]
I'll merge this time.

**Comment by sarutak** [comment]
@iRakson Could you update the screenshots in the description?

**Comment by iRakson** [comment]
> @iRakson Could you update the screenshots in the description?

Updated.

**Comment by iRakson** [comment]
Thank You. :) @sarutak @srowen

**Comment by sarutak** [comment]
Merged to `master`.

**Comment by sarutak** [comment]
@iRakson 
This PR passed PR builder's test but doesn't pass QA so I've reverted.

https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-2.7-hive-1.2/lastCompletedBuild/testReport/org.apache.spark.streaming/UISeleniumSuite/attaching_and_detaching_a_Streaming_tab/
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-2.7-hive-2.3/lastCompletedBuild/testReport/org.apache.spark.streaming/UISeleniumSuite/attaching_and_detaching_a_Streaming_tab/
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-2.7-hive-2.3-jdk-11/lastCompletedBuild/testReport/org.apache.spark.streaming/UISeleniumSuite/attaching_and_detaching_a_Streaming_tab/
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-3.2-hive-2.3/lastCompletedBuild/testReport/
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-3.2-hive-2.3-jdk-11/lastCompletedBuild/testReport/org.apache.spark.streaming/UISeleniumSuite/attaching_and_detaching_a_Streaming_tab/
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-2.7-hive-1.2/lastCompletedBuild/testReport/org.apache.spark.streaming/UISeleniumSuite/attaching_and_detaching_a_Streaming_tab/
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-sbt-hadoop-2.7-hive-2.3/lastCompletedBuild/testReport/

Could you confirm them?

**Comment by sarutak** [comment]
The suite pass on my laptop with both sbt and Maven so the suite can be flaky.

File: streaming/src/test/scala/org/apache/spark/streaming/UISeleniumSuite.scala [hunk 4]
```diff
@@ -125,24 +125,47 @@ class UISeleniumSuite
 
         // Check batch tables
         val h4Text = findAll(cssSelector("h4")).map(_.text).toSeq
-        h4Text.exists(_.matches("Active Batches \\(\\d+\\)")) should be (true)
+        h4Text.exists(_.matches("Running Batches \\(\\d+\\)")) should be (true)
```
Comment (iRakson) [inline streaming/src/test/scala/org/apache/spark/streaming/UISeleniumSuite.scala:128]: This is causing all the failures. I will remove these tests and raise again

**Pull request closed without merging**